proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,626 @@
+value structure (`TLV`) format and computes the required
+`TransactionHeaderParameter`s for the transaction. The `TLV` structure is well
+defined for accounts in solana program library repository we will extend the
+definition for instruction data. As we have multiple transaction header
+parameters we convert instruction data into a list of
+`TransactionHeaderParameter`. In case we fail to deserialize the transaction
+data, the transaction will fail with an error `InvalidInstructionData`. If same
+`TransactionHeaderParameter` is used multiple times or has multiple instructions
+using `TransactionHeaderProgram` we return an error `DuplicateInstruction`.
+Similarly, if we use old Compute Budget instructions with the new instruction to
+set the same parameters we return an error `DuplicateInstruction`.
+
+Program Id for the program `TransactionHeaderProgram`:
+
+```
+TrasactionHeader111111111111111111111111111
+```
+
+Definition for transaction header parameters:
+
+```
+enum TransactionHeaderParameter {
+  // existing compute budget parameters
+  RequestHeapFrame(u32),
+  SetComputeUnitLimit(u32),
+  SetComputeUnitPrice(u64),
+  SetLoadedAccountsDataSizeLimit(u32),
+  // Limit on maximum of PRAW fees to be paid
+  // Fees is set in microlamports per CU
+  SetProgramRebatableAccountWriteFees(u64),
+}
+```
+
+When the cluster receives a new transaction, with `TransactionHeaderParameter`
+instruction is decoded to get the `SetProgramRebatableAccountWriteFees` and then
+calculates the total fees required for the transaction message. Then we verify
+that the fee-payer has a minimum balance of: `per-transaction base fees` +
+`prioritization fees` +
+`Microlamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+Where PRAW fees is :
+`Microlamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+If the payer does not have enough balance, the transaction is not scheduled and
+fails with an error `InsufficientFunds`. Consider this case the same as if the
+payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+PRAW fees. For all accounts with PRAW fees, the fees paid should be greater than
+the required fees. In case of overpayment, the difference is stored in a
+variable and will be paid back to the payer in any case. In case the PRAW fees
+are insufficiently paid or not paid then the transaction fails with an error
+`ProgramRebatableAccountWriteFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The PRAW fees minus rebates and minus overpayment are transferred to the
+respective accounts lamport balance, from where the owner can collect them and
+then the rest would be transferred back to the payer.
+
+### Syscall to change PRAW fees
```

Comment:
how about explicitly mentioning that all input `address`es must be one of `AccountInfo`s given at program's entrypoint with is_writable=true; the getter might be okay with is_writable=false.....

Reply:
I think this is understood. Changing an account needs to pass `is_writable = true` at the program's entrypoint. Just trying to avoid unnecessary details as proposal is already too long.